mqtt.go

```diff
@@ -499,6 +516,124 @@ func (mc *mqttConn) writeBulkData(b *DeviceSyncedBulkData) error {
 	return errors.New("did not receive ack packet until timeout and bulk data dropped")
 }
 
+func (mc *mqttConn) sendBulkDataRequest(b *DeviceBulkDataRequest) error {
+	var qos byte
+
+	switch b.qos {
+	case QOSAtMostOnce:
+		qos = packet.QOSAtMostOnce
+	case QOSAtLeastOnce:
+		qos = packet.QOSAtLeastOnce
+	default:
+		return errors.New("unsupported qos level")
+	}
+
+	payload, _ := json.Marshal(b)
+
+	p := packet.NewPublishPacket()
+	p.PacketID = mc.getPacketID()
+	p.Message = packet.Message{
+		Topic:   fmt.Sprintf("/devices/%d/bulkData/%s/request", mc.dc.DeviceID, b.StreamID),
+		QOS:     qos,
+		Payload: payload,
+	}
+
+	mc.outPacket <- p
+
+	logInfo("[MQTT] bulkData request delivery for packet ID %d and requestID %s", p.PacketID, b.RequestID)
+
+	if b.qos == QOSAtMostOnce {
+		return nil
+	}
+
+	logInfo("[MQTT] waiting for PUBACK for packet ID %d", p.PacketID)
+
+	select {
+	case <-time.After(deviceEventRetryInterval):
+		msg := fmt.Sprintf("did not receive PUBACK of event delivery for packet ID %d", p.PacketID)
+		logError("[MQTT] %s", msg)
+		mc.reportError(fmt.Errorf("event delivery timeout: %s", msg))
+
+	case ack := <-mc.puback:
+		if ack.PacketID == p.PacketID {
+			logInfo("[MQTT] received PUBACK for packet ID %d", ack.PacketID)
+			return nil
+		}
+
+		msg := fmt.Sprintf(" received packet that does not match the Ack packet ID (%d) and the expected packet ID (%d).",
+			ack.PacketID, p.PacketID)
+		logInfo("[MQTT] %s", msg)
+		mc.reportError(fmt.Errorf("event delivery unmatched Ack Packet: %s", msg))
+	}
+
+	return errors.New("event dropped")
+}
+
+func (mc *mqttConn) sendSubscribeBulkDataResponse(sub *DeviceSubscribeBulkDataResponse) error {
+	p := packet.NewSubscribePacket()
+	p.PacketID = mc.getPacketID()
+
+	topic := fmt.Sprintf("/devices/%d/bulkData/%s/response", mc.dc.DeviceID, sub.StreamID)
+
+	logInfo("[MQTT] subscribing to BulkData response topic %s", topic)
+	p.Subscriptions = []packet.Subscription{
+		packet.Subscription{
+			Topic: topic,
+			QOS:   packet.QOSAtMostOnce, // MODE only supports QoS0 for subscriptions
+		},
+	}
+
+	mc.outPacket <- p
+
+	select {
+	case <-time.After(deviceEventRetryInterval):
+		msg := fmt.Sprintf("did not receive SUBACK of event delivery for packet ID %d", p.PacketID)
+		logError("[MQTT] %s", msg)
+		mc.reportError(fmt.Errorf("event delivery timeout: %s", msg))
+
+	case ack := <-mc.suback:
+		if ack.PacketID == p.PacketID {
+			logInfo("[MQTT] received SUBACK for packet ID %d", ack.PacketID)
+
+			if len(ack.ReturnCodes) != 1 {
```

Comment:
Since this is just a special version of the other suback code (L168), can you make it a function to reuse the code?

(Since the code won't live long, I think you should only spend 5 minutes on this suggestion. If it doesn't work, ignore it.)

Reply:
thanks! improved. https://github.com/moderepo/device-sdk-go/pull/28/commits/67d58a7a973392c538eabcd2d21159e84cd6b2b1